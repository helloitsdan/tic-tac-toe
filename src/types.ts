export enum BoardPosition {
  TOP_LEFT,
  TOP_CENTRE,
  TOP_RIGHT,
  MIDDLE_LEFT,
  MIDDLE_CENTRE,
  MIDDLE_RIGHT,
  BOTTOM_LEFT,
  BOTTOM_CENTRE,
  BOTTOM_RIGHT,
}

export enum PlayerPiece {
  PLAYER_ONE = 'PLAYER_ONE',
  PLAYER_TWO = 'PLAYER_TWO',
}

export enum MatchName {
  TOP_ROW = 'TOP_ROW',
  MIDDLE_ROW = 'MIDDLE_ROW',
  BOTTOM_ROW = 'BOTTOM_ROW',
  LEFT_COLUMN = 'LEFT_COLUMN',
  CENTRE_COLUMN = 'CENTRE_COLUMN',
  RIGHT_COLUMN = 'RIGHT_COLUMN',
  DIAGONAL_LEFT_TO_RIGHT = 'DIAGONAL_LEFT_TO_RIGHT',
  DIAGONAL_RIGHT_TO_LEFT = 'DIAGONAL_RIGHT_TO_LEFT',
}

export interface PlayerType {
  piece: PlayerPiece
  name: string
}

export type MatchTuple = [MatchName, BoardPosition[]]

export interface PlayerMatchType {
  match: MatchTuple
  piece: PlayerPiece
}

export type GameMovesType = Record<PlayerPiece, BoardPosition[]>

export interface GameStateType {
  players: PlayerType[]
  moves: GameMovesType
  winner: PlayerMatchType
}
