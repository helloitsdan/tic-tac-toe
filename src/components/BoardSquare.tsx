import React, { FunctionComponent } from 'react'
import classnames from 'classnames'

import { PlayerId } from '../types'
import PlayerIcon from './PlayerIcon'

interface BoardSquareProps {
  player: PlayerId
  onSelect: () => void
}

const BoardSquare: FunctionComponent<BoardSquareProps> = ({ player, onSelect }) => (
  <button
    className={classnames('c-board-button', {
      'u-player--player-one': player === PlayerId.PLAYER_ONE,
      'u-player--player-two': player === PlayerId.PLAYER_TWO,
    })}
    onClick={onSelect}
  >
    <PlayerIcon player={player} />
  </button>
)

export default BoardSquare
