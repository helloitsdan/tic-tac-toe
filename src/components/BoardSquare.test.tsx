import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import BoardSquare from './BoardSquare'
import { PlayerId } from '../types'

describe('BoardSquare', () => {
  it('renders', () => {
    const props = {
      player: PlayerId.PLAYER_ONE,
      onSelect: jest.fn(),
    }

    const { asFragment } = render(<BoardSquare {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('calls onSelect when clicked', () => {
    const props = {
      player: PlayerId.PLAYER_ONE,
      onSelect: jest.fn(),
    }

    const { getByRole } = render(<BoardSquare {...props} />)
    fireEvent.click(getByRole('button'))

    expect(props.onSelect).toHaveBeenCalled()
  })
})
