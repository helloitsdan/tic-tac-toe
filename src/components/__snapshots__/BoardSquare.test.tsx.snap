// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardSquare renders 1`] = `
<DocumentFragment>
  <button
    class="c-board-button u-player--player-one"
  >
    <svg
      class="c-player-piece"
      title="X"
    >
      X.svg
    </svg>
  </button>
</DocumentFragment>
`;
